crates/boojum/src/gadgets/tower_extension/algebraic_torus.rs

```diff
@@ -0,0 +1,494 @@
+        // Suppose r = f(g,i) / (f(w,i) * w^{-1}). Then, we require:
+        // f(g, i) = f(w, i) * (w^{-1}) * r
+        // Notice that `f(w,i)*w^{-1}` must yield an element
+        // from Fq6. Thus, we need one frobenius map + mul over Fq6, and
+        // one frobenius map + mul over Fq12.
+        let params = self.encoding.get_params();
+        let mut encoding_new = Fq6::allocate_from_witness(cs, witness_frob, params);
+
+        // rhs = f(w, i) * (w^{-1}) * r
+        // First, allocating the w^{-1}
+        let w_inverse = P::get_w_inverse_coeffs_c5();
+        let mut w_inverse: Fq2<_, _, _, <P::Ex6 as Extension6Params<T>>::Ex2> =
+            Fq2::constant(cs, w_inverse, params);
+
+        let mut rhs: Fq12<F, T, NonNativeFieldOverU16<F, T, N>, P> =
+            Fq12::one_imaginary(cs, params);
+        rhs = rhs.frobenius_map(cs, power);
+        rhs = rhs.mul_by_c5(cs, &mut w_inverse);
+
+        // Asserting that c1 is zero since rhs must be a pure Fq6 element at this point.
+        let boolean_true = Boolean::allocated_constant(cs, true);
+        let c1_is_zero = rhs.c1.is_zero(cs);
+        Boolean::enforce_equal(cs, &c1_is_zero, &boolean_true);
+        let mut rhs = rhs.c0.clone();
+
+        // Finishing rhs by multiplying by result
+        rhs = rhs.mul(cs, &mut encoding_new);
+
+        // lhs = f(g, i)
+        let mut lhs = self.encoding.clone();
+        lhs = lhs.frobenius_map(cs, power);
+
+        // Asserting that lhs == rhs
+        Fq6::enforce_equal(cs, &lhs, &rhs);
+
+        Self::new(encoding_new)
+    }
+
+    /// Computes the product of two Torus elements using the formula
+    ///
+    /// `(g, g') -> (g * g' + \gamma) / (g + g')`
+    ///
+    /// The formula handles the exceptional case when `g + g'` is zero.
+    pub fn mul<CS>(&mut self, cs: &mut CS, other: &mut Self) -> Self
+    where
+        CS: ConstraintSystem<F>,
+    {
+        // We compute multiplication unconstrained:
+        let witness_self = self.encoding_to_witness(cs);
+        let witness_other = other.encoding_to_witness(cs);
+        let witness_mul = P::torus_mul(witness_self, witness_other);
+
+        // Now, we constraint the multiplication with a cheaper version:
+        // g'' = (g * g' + \gamma) / (g + g') is equivalent to
+        // g'' * (g + g') = (g * g' + \gamma)
+        // Here, g'' is the new encoding.
+        let params = self.encoding.get_params();
+        let encoding_new = Fq6::allocate_from_witness(cs, witness_mul, params);
+
+        // lhs = g'' * (g + g')
+        let mut sum = self.encoding.clone().add(cs, &mut other.encoding);
+        let lhs = encoding_new.clone().mul(cs, &mut sum);
+
+        // rhs = {(g + g') == 0} ? zero : (g * g' + \gamma)
+        let mut gamma = Fq6::gamma(cs, params);
+        let mut rhs = self.encoding.clone().mul(cs, &mut other.encoding);
+        let rhs = rhs.add(cs, &mut gamma);
+
+        let zero = Fq6::zero(cs, params);
+        let is_zero_sum = sum.is_zero(cs);
```

Comment:
I assume you got this code from our old version of franklin-crypto. There are two versions, safe and not. This code was taken from an unsafe version. The point is that we do not check that the sum of g1 and g2 is zero because the constraint guarantees that malicious will not find such g2 = -g1 that won't be the root of gamma. The unsafe version was made on the assumption to reduce constraints ( that means that we are pretty sure that this is not underconstrained)

Reply:
Indeed, thanks for pointing that out!